src/lib/utils/ct_utils.h

```diff
@@ -96,30 +96,36 @@ inline bool poison_has_effect() {
  * Poison a single integral object
  */
 template <std::integral T>
-constexpr void poison(T& p) {
+constexpr T& poison(T& p) {
    poison(&p, 1);
+   return p;
 }
 
 template <std::integral T>
-constexpr void unpoison(T& p) {
+constexpr T& unpoison(T& p) {
    unpoison(&p, 1);
+   return p;
 }
 
 /**
  * Poison a contiguous buffer of trivial objects (e.g. integers and such)
  */
 template <ranges::spanable_range R>
-   requires std::is_trivially_copyable_v<std::ranges::range_value_t<R>>
-constexpr void poison(R&& r) {
+   requires std::is_trivially_copyable_v<std::ranges::range_value_t<R>> &&
+            (!requires(const R& x) { x._const_time_poison(); })
+constexpr R&& poison(R&& r) {
    std::span s{r};
    poison(s.data(), s.size());
+   return std::forward<R>(r);
 }
 
 template <ranges::spanable_range R>
-   requires std::is_trivially_copyable_v<std::ranges::range_value_t<R>>
-constexpr void unpoison(R&& r) {
+   requires std::is_trivially_copyable_v<std::ranges::range_value_t<R>> &&
+            (!requires(const R& x) { x._const_time_unpoison(); })
+constexpr R&& unpoison(R&& r) {
```

Comment:
Here and in the other overloads: Please use `decltype(auto)` as return value. The `std::forward<R>()` will "do the right thing", by either returning as an rvalue or lvalue reference. But I'm not 100% sure that the explicit return type `R&&` always resolves to the correct reference type. Using `decltype(auto)` let's the compiler figure it out from the return value of `std::forward`.

```suggestion
constexpr decltype(auto) unpoison(R&& r) {
```

This might even obsolete the `std::forward()` in the return statement. I'll have to ponder over this.

Reply:
I bow to the master :D I tried using `auto`, which did not handle references correctly. But indeed, `decltype(auto)` does the right thing. I admit I didn't know there was a difference between both variants.